Fix race condition on cache directory creation

moio opened the pull request:
### What does this PR do?

`fileclient.py`: a race condition will result in a stack trace.

Details: in general `os.makedirs()` will raise `OSError` in case the directory passed as argument
already exists. What we do in `_cache_loc` is to check for directory existence a few instructions before the
call, which leaves a tiny time window in which the directory might
be concurrently created between the `isdir()` check and the `makedirs()` call, which can result in the race condition.

In some unlucky cases under heavy I/O the following stack trace is produced:

The minion function caused an exception: Traceback (most recent call last):
...
  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 165, in cache_file
    return self.get_url(path, '', True, saltenv, cachedir=cachedir)
...
  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 126, in _cache_loc
    os.makedirs(destdir)
  File "/usr/lib64/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: <PATH>

I can reliably reproduce this with `minionswarm.py` in an AWS install.

This PR contains a fix for that problem.

### What issues does this PR fix or reference?

The issue described above, I can open one if needed.

### Previous Behavior

Stack trace above.

### New Behavior

No stack trace.

### Tests written?

No.

moio committed:
Fix race condition on cache directory creation

os.makedirs() will raise OSError in case the directory passed as argument
already exists. We do check that this is not the case right before the
call, but there is still a tiny time window in which the directory might
be concurrently created between the isdir() check and the makedirs() call.

In some unlucky cases under heavy I/O the following stack trace is produced:

The minion function caused an exception: Traceback (most recent call last):
...
  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 165, in cache_file
    return self.get_url(path, '', True, saltenv, cachedir=cachedir)
...
  File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 126, in _cache_loc
    os.makedirs(destdir)
  File "/usr/lib64/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: <PATH>

moio commented:
@isbm @dincamihai comments appreciated, as this is my first PR against Salt core!

isbm commented on salt/fileclient.py:
```diff
@@ -142,12 +142,18 @@ def _cache_loc(self, path, saltenv='base', cachedir=None):
+        try:
             os.makedirs(destdir)
+        except OSError as e:
+            pass
```
I would still quack about that into logs why this had happened. I still believe that muting all that completely isn't best, because you might have another reason why that had happened and _maybe_ you want to see it. So I would:
```python
try:
    os.makedirs(destdir)
except OSError as err:
    e_code, e_msg = err.args
    if e_code != 17:  # Other than already exists
        log.error("Unable to create a directory: '{message}' (exit code: {code})".format(
            message=e_msg, code=e_code)
```

sergeizv replied:
I would prefer errno.EEXIST versus magic numbers :)

s0undt3ch replied:
`errno` is also more portable.

isbm reviewed (changes requested):
@moio maybe you want to consider this change. Thanks!

moio commented:
Sure @isbm, totally valid point. I changed the code to work just as before for non-EEXIST cases. Feedback from @sergeizv and @s0undt3ch also incorporated, thank you very much!

isbm commented:
@moio thanks! @cachedout seems like we're good to go. :smile:

rallytime commented:
Go Go Jenkins!

cachedout commented:
Go Go Jenkins!

isbm commented on salt/fileclient.py:
```diff
@@ -142,12 +143,21 @@ def _cache_loc(self, path, saltenv='base', cachedir=None):
+            if exc.errno == errno.EEXIST:  # already there, move on
+                pass
+            else:
+                raise
```
What is wrong with:
```python
try:
    os.makedirs(destdir)
except OSError as exc:
    if exc.errno != errno.EEXIST:
        raise

moio replied:
Absolutely nothing, I only find my version more legible, because it's more explicit in what we whitelist - yours is more concise OTOH.

FMPOV this is just a matter of taste at this point and I am fine either way.

isbm replied:
The `exc.errno != errno.EEXIST` _is_ a whitelist, no? Expands with:
```python
if exc.errno not in WHITELIST:
    raise
```

moio replied:
@cachedout any style advice from you to move this forward?

isbm replied:
@moio you're here asking advice whether you should or not write:
```python
if trouble:
  pass
else:
  do_something()
```
We normally write:
```python
if not trouble:
  do_something()
```
I fail to understand why these two extra lines needs to be there as a noise in the code.

cachedout replied:
I personally tend to write it like @isbm has outlined but we do not have a defined style for this.

moio replied:
Changed, as consistency also has value.

@isbm can you approve the review now?

cachedout commented:
Go Go Jenkins!

isbm reviewed (approved):
@moio absolutely! :heart:

Commit:
Extra space for lint compliance
